src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,241 @@
+#pragma once
+#include <chrono>
+#include <thread>
+#include <condition_variable>
+#include "rgw_common.h"
+
+#if !defined(dout_subsys)
+#define dout_subsys ceph_subsys_rgw
+#define def_dout_subsys
+#endif
+
+#define dout_context g_ceph_context
+
+class QosDatastruct {
+  enum {
+    RGW_QOS_OPS,
+    RGW_QOS_BW,
+    RGW_QOS_ROUND_OPS,
+    RGW_QOS_ROUND_BYTES
+  };
+  struct qos_entry {
+    // 3 is the enum size
+    std::atomic_int64_t read[5] = {0};
+    std::atomic_int64_t write[5] = {0};
+    int64_t ts = ceph::coarse_real_clock::now().time_since_epoch().count();
+    bool first_run = true;
+    std::mutex ts_lock;
+  };
+  const size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::mutex insert_lock;
+  std::condition_variable *cv;
+  std::atomic_bool *replacing;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  bool is_read_op(const char *method) {
+    if((strcmp(method,"GET") == 0) || (strcmp(method,"HEAD") == 0))
+      return true;
+    return false;
+  }
+    // find or create an entry, and return its iterator
+  auto find_or_create(const std::string& key) {
+    if(qos_entries.size() > 0.9 * map_size && *replacing == false)
+    {
+        *replacing = true;
+        cv->notify_all();
+    }
+    auto ret = qos_entries.find(key);
+    if(ret == qos_entries.end())
+    {
+      insert_lock.lock();
+      ret = qos_entries.emplace(std::piecewise_construct,
+                                 std::forward_as_tuple(key),
+                                 std::forward_as_tuple()).first;
+      insert_lock.unlock();
+    }
+    return ret;
+  }
+
+  bool increase_entry_read(hash_map::iterator& it, int64_t ops_limit, int64_t bw_limit) {
+    int64_t ops = it->second.read[RGW_QOS_OPS];
+    int64_t bw = it->second.read[RGW_QOS_BW];
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((ops - 1 < 0) && (ops_limit > 0)) ||
+      (bw < 0 && bw_limit > 0))
+      return true;
+    // we don't want to reduce ops' tokens if we've rejected it.
+    --it->second.read[RGW_QOS_OPS];
+    return false;
+  }
+
+  bool increase_tokens(hash_map::iterator& it, int64_t curr_timestamp,
+                       const RGWQoSInfo& tenant)
+  {
+    constexpr int fixed_float = 10000;
+    it->second.ts_lock.lock();
+    if(curr_timestamp > it->second.ts)
+    {
+      double time_in_ns = (double)(curr_timestamp - it->second.ts) / std::nano::den / 60.0; // / 60 to make it work with 1 min token bucket
+      dout(0) << "time_in_ns " << time_in_ns << dendl;
+      it->second.ts = curr_timestamp;
+
+      it->second.write[RGW_QOS_ROUND_OPS] += static_cast<int64_t>(tenant.max_write_ops * time_in_ns * fixed_float) % fixed_float;
+      int round = it->second.write[RGW_QOS_ROUND_OPS] / fixed_float;
+      it->second.write[RGW_QOS_ROUND_OPS] = it->second.write[RGW_QOS_ROUND_OPS] - round * fixed_float;
```

Comment:
instead of tracking `RGW_QOS_ROUND_OPS/BYTES` separately, what do you think about redefining the `RGW_QOS_OPS` and `RGW_QOS_BW` counters to be in multiples of 10000?

that would just mean multiplying `tenant.max_write_ops/bytes` by 10000 for comparisons, and dividing the counters by 10000 when you want their actual values:
```c++
int64_t qos_entry::read_ops() const { return read[RGW_QOS_OPS] / 10000; }
int64_t qos_entry::write_ops() const { return write[RGW_QOS_OPS] / 10000; }
```
this way, we only need the two counters (RGW_QOS_OPS/BW), and this math in `increase_tokens()` is a lot simpler